solutions/cf-1418C.mdx

```diff
@@ -61,20 +61,68 @@ int main() {
 		 */
 		dp[1][0] = 0;
 
-		for (int i = 0; i < n; i++) {
+		for (int j = 0; j < n; j++) {
 			// the opposite player switches on the previous move.
-			dp[0][i + 1] = min(dp[0][i + 1], dp[1][i] + bosses[i]);
-			dp[1][i + 1] = min(dp[1][i + 1], dp[0][i]);
+			dp[0][j + 1] = min(dp[0][j + 1], dp[1][j] + hard[j]);
+			dp[1][j + 1] = min(dp[1][j + 1], dp[0][j]);
 
 			// the opposite player switches from the previous two moves.
-			if (i + 2 <= n) {
-				dp[0][i + 2] = min(dp[0][i + 2], dp[1][i] + bosses[i] + bosses[i + 1]);
-				dp[1][i + 2] = min(dp[1][i + 2], dp[0][i]);
+			if (j + 2 <= n) {
+				dp[0][j + 2] = min(dp[0][j + 2], dp[1][j] + hard[j] + hard[j + 1]);
+				dp[1][j + 2] = min(dp[1][j + 2], dp[0][j]);
 			}
 		}
 		cout << min(dp[0][n], dp[1][n]) << endl;
 	}
 }
 ```
 </CPPSection>
+
+<JavaSection>
+```java
+import java.io.*;
+import java.util.Arrays;
+
+public class MortalKombatTower {
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		int t = Integer.parseInt(br.readLine());
+
+		for (int i = 0; i < t; i++) {
+			int n = Integer.parseInt(br.readLine());
+			int[] hard = Arrays.stream(br.readLine().split(" "))
+					.mapToInt(Integer::parseInt)
+					.toArray();
+
+			/*
+			 * dp[i][j] = the min amount of skip points needed if it is
+			 * player i's turn (you are 0 and your friend 1) on the j-th boss.
+			 * At the beginning (before fighting any bosses), dp[1][0] = 0.
+			 */
+			int[][] dp = new int[2][n + 1];
+			for (int j = 0; j <= n; j++) {
+				dp[0][j] = 1000000000;
+				dp[1][j] = 1000000000;
```

Comment:
`Integer.MAX_VALUE`

Reply:
will lead to overflow. I can add a constant INTMAX=1000000000 if that's better.

Reply:
if you're doing overflow just check if you're adding INT MAX

Reply:
sry u mean i *should* use the constant?

Reply:
no i mean
use `Integer.MAX_VALUE`
you shouldn't really be adding those invalid values anyways, so it's best to add a check

Reply:
So i should do 
```
dp[0][j + 2] = Math.min(dp[0][j + 2], dp[1][j] == Integer.MAX_VALUE ? Integer.MAX_VALUE :  dp[1][j] + bosses[j] + bosses[j + 1]);
```
?

Reply:
if it's that long i'd just use an if statement

Reply:
is this acceptable `dp[0][j] = Integer.MAX_VALUE - 3` ?

At this point, I *don't* think this is more clear but if it's preferred..
```
if (dp[1][j] < Integer.MAX_VALUE) {
    dp[0][j + 2] = Math.min(dp[0][j + 2], dp[1][j] + bosses[j] + bosses[j + 1]);
}
```

Reply:
why the -3?
if it's MAX, you shouldn't be adding it in the first place, no?

Reply:
actually, mathematically, adding a number to $\infty$ still lead to infinity. In our case, infinity means that it is infeasible to use this state (we need infinity points), and therefore the next state based on this should also be infeasible, i.e. valued with infinity. We actually don't really want the MAX VALUE of a data type but rather the idea of infinity. Using if statement is one way to simulate this (adding to inf is inf), or, use a sightly smaller val than MAX VALUE, which makes it to *behave* like inf since one can add to it and it's still larger than a certain line, from which we call a number infinity. 

As said, I don't think adding two if statements (add thereby more nesting levels) make the code any easier to read. But if u insist, I don't really have a big problem with that either.

Reply:
then set it to infinity if one of the values is infinity?

Reply:
so `static final int INF = 1000000000;`?

Reply:
no, just use `Integer.MAX_VALUE` and do checks

Reply:
alright will do